components/kyma-environment-broker/internal/process/upgrade_kyma/initialisation.go

```diff
@@ -232,13 +245,28 @@ func (s *InitialisationStep) checkRuntimeStatus(operation internal.UpgradeKymaOp
 		msg = *status.Message
 	}
 
+	// execute pre- steps
+	operation, delay, err := s.performRuntimeTasks(UpgradePreSteps, operation, instance, log)
```

Comment:
The problem with this naming is that this is not a "pre" step, it's executed right after the upgrade has been started.
I would just not use this additional layer of abstraction to try to make it more generic, it increases code complexity (extra levels of function calls and if conditions), without bringing too much benefit.

I would just fall back to ensure AVS evaluations being in maintenance  by using `em.SetMaintenanceStatus()` / `em.RestoreStatus()`  directly when `status.State` is not terminal, and AVS evaluations being restored when they are. 
The logic to return prematurely when there is AVS error even if the provisioning status is Succeeded is good as it won't termonate this KEB operations and will allow retry to ensure AVS status. 

With this you could extract/eliminate functions `executePreUpgradeTasks`, `executePostUpgradeTasks` and `performRuntimeTasks`, saving two levels of function calls, and increasing readability a LOT.

Reply:
The initial approach was focused on extending the tasks. As mentioned, it doesn't do a good job to verify calls as it can retrigger the same flow (`executePreUpgradeTasks`, `executePostUpgradeTasks`). 

This has been replaced with init and finish steps that will be triggered in `performRuntimeTasks` as that should be an independent handler of all Kyma Upgrade vital dependency calls at some stage of execution.